### PR TITLE
Support timeouts for lucene search provider

### PR DESCRIPTION
**Pull request by aklish**
…ger'

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 1]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
...
+            public AccessibleTimeLimitingCollector newCollector() throws IOException {
+                return new AccessibleTimeLimitingCollector(
+                        TopScoreDocCollector.create(perPage, lastEntry), Counter.newCounter(false), searchTimeout
+                );
+            }
+
+            @Override
+            public TopDocs reduce(Collection<AccessibleTimeLimitingCollector> collectors) throws IOException {
+                final TopDocs[] docs = collectors.stream()
+                        .map((collector) -> collector.getWrappedCollector().topDocs())
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Can be decomposed into two method reference-based mappings:

```java
.map(AccessibleTimeLimitingCollector::getWrappedCollector)
.map(TopDocsCollector::topDocs)
```

Decomposing allows for more flexible streams and reduces complexity by making each step simpler.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 2]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
...
+            @Override
+            public AccessibleTimeLimitingCollector newCollector() throws IOException {
+                return new AccessibleTimeLimitingCollector(
+                        TopScoreDocCollector.create(perPage, lastEntry), Counter.newCounter(false), searchTimeout
+                );
+            }
+
+            @Override
+            public TopDocs reduce(Collection<AccessibleTimeLimitingCollector> collectors) throws IOException {
+                final TopDocs[] docs = collectors.stream()
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: `final` is not needed here

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 3]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
             int perPage,
             int currentPage
     ) {
+        /**
+         * Works around the lack of a way to get at the underlying Collector in TimeLimitingCollector
+         */
+        class AccessibleTimeLimitingCollector extends TimeLimitingCollector {
+            private TopScoreDocCollector wrappedCollector;
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Would be nice to make this `final` to better provide immutability where we can.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 4]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
...
+             */
+            public TopScoreDocCollector getWrappedCollector() {
+                return wrappedCollector;
+            }
+        }
+
+        /**
+         * This parallelizes the lucene search - each thread using a time limited collector.
+         */
+        final CollectorManager<AccessibleTimeLimitingCollector, TopDocs> manager
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: `final` not needed here.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 5]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
...
+            @Override
+            public TopDocs reduce(Collection<AccessibleTimeLimitingCollector> collectors) throws IOException {
+                final TopDocs[] docs = collectors.stream()
+                        .map((collector) -> collector.getWrappedCollector().topDocs())
+                        .collect(Collectors.toList())
+                        .toArray(new TopDocs[collectors.size()]);
+
+                return TopDocs.merge(perPage, docs);
+            }
+        };
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: It would be great to pull this anonymous class (and the helper wrapper class) out into it's own stand-alone class, for a number of reasons.

1. I think the capability is generally useful, especially if another part of the application needs to deal with Lucene in a different way. Having time-bounded searches is generally a good idea
2. Having these classes defined in-line in the method will incur more performance overhead, since a new class is generated every method call, rather than in a more static manner that will happen once per class-load
3. The classes are very hard to unit-test when they are embedded in the method. Making them top-level makes them much easier to test on their own as a small, focused unit with little environment / setup overhead
4. The `LuceneSearchProvider` is already rather large. Adding more responsibilities to it adds more complexity, but having this responsibility (ie. time-bounding TopDoc searches) encapsulated into it's own top-level class keeps the complexity better in-check

It probably makes sense to have a single top-level `TimeLimitingTopDocsCollectorManager` class, with the `AccessibleTimeLimitingCollector` as an inner-class, since it's pretty tightly associated with time-limiting top docs.

I would also leave the `searchTimeout` in the `LuceneSearchProvider`, and pass it to the `TimeLimitingTopDocsCollectorManager` as a constructor param, since we've generally found it better to keep the interactions with `SystemConfig` as high in the stack as we can to reduce coupling and keep responsibilities small for pieces low in the stack.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 6]
```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
...
+
         lock.readLock().lock();
         try {
-            return lastEntry == null ?
-                    indexSearcher.search(query, perPage) :
-                    indexSearcher.searchAfter(lastEntry, query, perPage);
+            return indexSearcher.search(query, manager);
```
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: It looks like we're killing pagination here. Is that intentional?

**Comment by archolewa** [comment]
After talking to @aklish off line, it looks like we're not (hopefully) breaking pagination, the pagination is just being moved into the collectors that are being built and merged.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 5 above)
Comment (aklish) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: I think it makes sense to pull out the non-anonymous class since it is named.  But I think the anonymous class makes more sense to leave inside the function.

Pulling out the anonymous class means it has to be named and it also pollutes the list of nouns/classes in the source tree (increasing the mental model of that package).

I think if someone else wants this functionality, it can be refactored at that time.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 6 above)
Comment (aklish) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: It is actually handled in the newCollector function.  That function constructs a collector with `lastEntry` as an argument.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 5 above)
Comment (aklish) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: One more point about this, the anonymous class is actually using closure to take arguments from the parent context.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 6 above)
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Yeah, I looked closely at this too, and the code we have here is almost exactly the same logic that the TopN collector is doing internally. The difference is that, by using a manager, we're actually parallelizing the work across more than 1 thread and the reducing down, which we were not doing before, so things _should_ be faster (and more concretely bounded).

The only worry I have is that having more disjoint thread pools (ie. Lucene will use it's own for this parallelism) makes it harder to tune things, since you can have more contention for processors more easily, but if we see that, we can deal with it.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/DataServlet.java [hunk 7]
```diff
@@ -85,6 +74,15 @@
...
 import javax.ws.rs.core.UriInfo;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import static com.yahoo.bard.webservice.web.handlers.workflow.DruidWorkflow.REQUEST_WORKFLOW_TIMER;
+import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
+import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
+import static javax.ws.rs.core.Response.Status.REQUEST_TIMEOUT;
```
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/DataServlet.java]: Looks like your import configuration doesn't agree with ours. The imports should be static, then `com.yahoo` then `com` then `org` then other, then `java`, then `javax`.

**Comment by archolewa** [comment]
A small nitpick: It'd be nice if we could add a bit of documentation to the `SearchProvider` interface that explains that the `getDimensionsPaged` method may throw the `TimeoutException`.  For one, it'll help people implementing future `SearchProviders` know that timing out is a thing they can do.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProviderManager.java [hunk 8]
```diff
@@ -7,11 +7,10 @@
 import com.yahoo.bard.webservice.util.Utils;
 import com.yahoo.bard.webservice.web.util.PaginationParameters;
 
+import javax.validation.constraints.NotNull;
 import java.util.HashMap;
 import java.util.Map;
 
-import javax.validation.constraints.NotNull;
-
```
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProviderManager.java]: Should probably undo this change as well.

**Comment by archolewa** [comment]
👍 Once the changes to the import orders that I pointed out have been addressed.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/DataServlet.java [hunk 9]
```diff
@@ -448,6 +446,9 @@ public void getData(
         } catch (NoMatchFoundException e) {
             LOG.info("Exception processing request", e);
             asyncResponse.resume(RequestHandlerUtils.makeErrorResponse(INTERNAL_SERVER_ERROR, e, writer));
+        } catch (TimeoutException e) {
+            LOG.info("Exception processing request", e);
+            asyncResponse.resume(RequestHandlerUtils.makeErrorResponse(REQUEST_TIMEOUT, e, writer));
```
Comment (aklish) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/DataServlet.java]: I think 408 is actually what we want here.  A 408:
```The server did not receive a complete request message within the time that it was prepared to wait```
